File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        HashSet<ListNode> li = new HashSet<>();
        ListNode a = headA;
        while (a!=null){
            li.add(a);
            a = a.next;
        }
        ListNode b = headB;
        while(b!=null){
            if(li.contains(b)){
                return b;
            }
            li.add(b);
            b=b.next;
        }
        return null;
    }
}

File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        a = headA
        s = set()
        while a:
            if a not in s:
                s.add(a)
            a = a.next
        b = headB
        while b:
            if b not in s:
                s.add(b)
            else:
                return b
            b = b.next
